doc/extract-snp.1
.TH extract-snp 1 "4 May 2016" "ProbABEL 0.5.0"
.SH NAME
extract-snp \- tool to extract dosage values for a given SNP for all or some individuals.
.SH SYNOPSIS
.B extract-snp
.B "-f\ \fIFILE\fR"
.B "-s\ \fISNPNAME\fR"
.RI "[" "OPTIONS" "]"
.SH DESCRIPTION
.B extract-snp
extracts the dosage for a SNP for one or more individuals from a file in filevector
format (.fvi/.fvd).
.PP
.SH OPTIONS
.SS Required command line options
.TP
.BI "-f, \-\^\-file" " FILE"
Name of the filevector file from which to extract the dosages. This
can be either the file with the .fvi or the .fvd extension.
.TP
.BI "-s, \-\^\-snp" " snpname"
Name of the SNP or variant for which to extract the dosages
.RE
.SS Other, non-mandatory options
.TP
.BI "-i, \-\^\-id" " individualID"
Extract the dosage for only one individual (with ID
\fIindividualID\fR) for the SNP listed with the \-\-snp option.
.RE
.TP
.BI "-I, \-\^\-listids"
List all ID names in the fv file. If this option is used, any other
options are ignored.
.RE
.TP
.BI "-L, \-\^\-listsnps"
List all SNP names in the fv file. If this option is used, any other
options are ignored.
.RE
.TP
.BI "-h, \-\^\-help"
Print usage information to the screen and exit.
.RE
.TP
.BI "-d, \-\^\-debug"
Turn on debug mode. This will print a lot of extra output to the
screen, for example it lists the names of all individuals and SNPs
found in the file.
.RE
.SH EXAMPLES
This extracts the dosages for all individuals and saves them in the
file dosages.txt:
.nf
.RS
extract-snp \-f ../checks/inputfiles/mmscore_gen.dose.fvi \-s rs70099 > dosages.txt
.RE
.fi
.SH "SEE ALSO"
palinear(1), palogist(1), pacoxph(1), probabel(1), and the R package DatABEL.
.SH BUGS
The bugtracker is located at
.br
https://github.com/GenABEL-Project/ProbABEL/issues
.SH AUTHORS
Lennart C. Karssen


doc/ChangeLog
***** v.0.x.0 ()
* ProbABEL can now read gzipped info files, map files, invsigma files and
  genotype data. This requires that the Boost IOStream library is
  installed when compiling. For clarity: reading from gzipped
  DatABEL/filevector files is not supported (as it would defeat the
  purpose of fast access to the data). This closes Issue #12 "Allow 
  reading of gzipped files". 
* Removed the requirement on having 7 columns in the info file. However,
  it is the user's responsibility to make sure the first 7 columns contain
  the right information.

***** v.0.5.0 (2016.05.04)
* ProbABEL now depends on the Eigen library. The option to compile without
  Eigen has been removed. Eigen 3.2.1 is included with ProbABEL. The
  directory src/eigen-3.2.1 contains header files we need. The full Eigen
  source code tar-ball is included as src/3.2.1.tar.bz2.
* Fixed Issue #15: Coxph interactions not working; when using the
  --interaction option the name of the model was wrong (e.g. for
  --interaction=1, the interaction term was shown as mu * SNP_A1, which
  obviously is not what we meant to do).
* Fixed Github issue #16: "Too many NANs for SNPs with low `Rsq` when some
  commandline options are specified"
* Added --flipmaf command line option to palinear, pacoxph and palogist.
  If this option is selected, ProbABEL will 'flip' the alleles A1 and A2
  for those variants for which the Freq1 value in the info file is less
  than 0.5. Changing the reference and effect allele in this way can
  sometimes (for some genetic models and low MAF), reduce regression
  errors and thus result in proper estimates which otherwise be set to
  NaN. Many thanks for the group of Prof. Kiemeney at the dept. of Health
  Evidence Radboud UMC in Nijmegen (NL) for sponsoring the work on this
  feature! Also thanks to Tessel Galesloot from the same group for
  extensive testing of this release. (Pull Request #24, also fixes bug #5).
* Don't set the betas and se_betas for all covariates in case Cox
  regression problems like singularities occur, but only the beta/se_beta
  of the corresponding covariate. Also make the corresponding warning
  messages more clear by including the name of the genetic model that is
  being tested.
* Not-a-number values in the output are now printed as "nan" instead of
  "NaN".
* Linear regression mmscore option is 2 times faster
* Speedup of Linear regression: (measured with multiple runs with 2 covariates,
  33815 SNP and 3485 people (using all) compared to v.0.4.3)
** Reading mldose/mlprob files 14 times faster
** Calculation of regression is more than 3.5 faster
** Overall runtime using above settings is 5 times faster
* Handles multiple variants of NaN (NA,Na,Nan,na,nan) correct while reading of
  mldose/mlprob files
* Fixed Issue #21: Remove "interaction_only" option. According to Yurii's
  post on the forum this option never worked as expected (Pull Request
  #23). Thanks to Matthias Wuttke for reminding us of this bug.
* Fixed bugs #5883 & #6010 (on R-forge): The main effect is displayed in
  the output with the  `interaction_only` option. Thanks to Maksim
  Struchalin & Lennart Karssen for fixing it and to Farid Radmanesh for
  reporting it. This was only a display error, computations were not
  affected.
* Fixed bug #5982: ProbABEL's make install fails on MacOS X and FreeBSD
  with sed error. Thanks to forum user mmold for reporting the bug.


***** v.0.4.5 (2015.05.26)
* Fixed bug #6054: "Not all ProbABEL's short command line options are
  correctly parsed."
* Fixed bug #6041: "ProbABEL's Cox module reports too many errors (beta
  may be infinite, setting beta and se to 'NaN')". Thanks to Anne
  Grotenhuis from the Radboud Medical Centre Nijmegen and forum user
  quentin300 for their extensive bug reports and help in testing the fix.
* Fixed bug #1266: "pacoxph with no covariates"; ProbABEL now also works
  when no covariates have been provided. Thanks to Aaron Joon for
  reporting this bug back in 2011 and to Anne Grotenhuis and to forum user
  quentin300 for pushing this bug to my attention and their help in beta
  testing.
* pacoxph now displays the regression equation correctly.
* Minor fix to the manual: the Rsq values in the info file should be >
  1e-16 (and not > 0 as mentioned before) otherwise the output will be set
  to 'nan'.
* Minor fix to the man-pages: according to the man-pages palogist,
  palinear and pacoxph all did regression using a linear model.


***** v.0.4.4 (2014.11.07)
* Fixed bug #5729 in the Cox PH module. Some checks for problems with the
  regression were incorrectly implemented. Thanks to Matthias Wuttke from
  the University Medical Centre Freiburg, Anne Grotenhuis from the Radboud
  Medical Centre Nijmegen, and Luba Pardo and Joris Verkouteren from the
  Erasmus Medical Centre Rotterdam for their time and effort in reporting
  the bug, helping to identify the problem and testing the fix.
* Backported fix of bug #5982: ProbABEL's make install fails on MacOS X
  and FreeBSD with sed error. Thanks to forum user mmold for reporting the bug.


***** v.0.4.3 (2014.04.01)
* Speed-up of a factor of ~ 2 for linear, logistic and Cox regression when
  using filevector input files.
* Fixed bug #5404: "ProbABEL's R check for Cox regression doesn't check if
  the survival package is installed".
* Fixed bug #5403: "The ProbABEL manual doesn't contain any information on
  how to install ProbABEL"


***** v.0.4.2 (2014.01.02)
* The 'probabel.pl' script is now simply renamed to 'probabel' (a user
  shouldn't care what scripting language we use). For at least several
  releases to come, the old script name will still exist (as a link to the
  original) and a warning message is displayed when the user runs the
  .pl script. This should give people time to adjust their pipelines.
* Fix bug #4919: Too small reading buffers for long alleles in mach info
  and legend files. Thanks to Daniel Taliun for reporting the bug and

src/extract-snp.cpp
/*******************************************************************************
 *
 * This program extracts a given row (dosages for all individuals for a given
 * SNP) from filevector files.
 *
 * (C) 2016 L.C. Karssen
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 ******************************************************************************/

#include <stdio.h>
#include <getopt.h>
#include <string>
#include <iostream>
#include <fstream>
#include <map>

#if HAVE_CONFIG_H
#include "config.h"
#endif

using std::cout;
using std::cerr;
using std::endl;

#include "fvlib/Logger.h"
#include "fvlib/FileVector.h"
#include "fvlib/CastUtils.h"


void info(char *program_name)
{
    cout << program_name
         << " extracts the dosage for a SNP for one or more individuals from a"
         << " file in filevector format (.fvi/.fvd)." << endl;
    cout << endl;
    cout << "Usage: " << program_name << " --file <fv file> --snp <snpname>"
         << endl;
    cout << "   or: " << program_name << " -f <fv file> -s <snpname>" << endl;
    cout << endl;
    cout << "Additional options:" << endl;
    cout << "\t--listsnps (-L); list all SNP names from the fv file"
         << " (the program exist after this)"
         << endl;
    cout << "\t--listids (-I); list all ID names from the fv file"
         << " (the program exist after this)"
         << endl;
    cout << "\t--id (-i) <idname>: "
         << "only print the dosage for the given SNP and ID" << endl;
    cout << "\t--debug (-d): show debugging output" << endl;
    cout << "\t--help (-h): show this information" << endl;
    cout << endl;
    cout << program_name << " is part of " << PACKAGE
         << " v" << PACKAGE_VERSION << endl;
    cout << "(C) 2015 Lennart C. Karssen, PolyΩmica, NL" << endl;
}


int main(int argc, char* argv[])
{
    if (argc < 2)
    {
        info(argv[0]);
        exit(3);
    }

    int next_option;
    const char * const short_options = "df:hi:ILs:";
    const struct option long_options [] =
        {
            {"debug",    0, NULL, 'd'},
            {"file",     1, NULL, 'f'},
            {"help",     0, NULL, 'h'},
            {"id",       1, NULL, 'i'},
            {"listids",  0, NULL, 'I'},
            {"listsnps", 0, NULL, 'L'},
            {"snp",      1, NULL, 's'},
            {NULL,       0, NULL, 0  }
        };

    char *program_name = argv[0];
    bool debug    = false;
    bool listSNPs = false;
    bool listIDs  = false;
    string inputFileName = "";
    string idname        = "";
    string snpname       = "";
    do
    {
        next_option = getopt_long(argc, argv,
                                  short_options, long_options, NULL);
        switch (next_option)
        {
        case 'h':
            info(program_name);
            exit(0);
        case 'd':
            debug = true;
            break;
        case 'f':
            inputFileName = string(optarg);
            break;
        case 'i':
            idname = string(optarg);
            break;
        case 'I':
            listIDs = true;
            break;
        case 'L':
            listSNPs = true;
            break;
        case 's':
            snpname = string(optarg);
            break;
        case '?': break;
        case -1 : break;
        }
    } while (next_option != -1);

    if (inputFileName.compare("") == 0)
    {
        cerr << "Error: No filevector file specified.\n";
        info(program_name);
        exit(4);
    }

    if (debug) cout << "Input file is '" << inputFileName << "'." << endl;


    FileVector fv(inputFileName, 64, true);

    if (listSNPs)
    {
        // Only print SNPs to stdout.
        for (unsigned long int row = 0; row < fv.getNumVariables(); row++)
        {
            string current_snpname = fv.readVariableName(row).name;
            cout << current_snpname << " ";
        }
        cout << endl;
        exit(0);
    }

    if (listIDs || debug)
    {
        if (debug)
        {
            cout << "---------- List of IDs ----------" << endl;
        }
        // Only print IDs to stdout.
        for (unsigned long int col = 0; col < fv.getNumObservations(); col++)
        {
            cout << fv.readObservationName(col).name << " ";
        }

        cout << endl;

        if (debug)
        {
            cout << "---------- End of IDs ----------" << endl;
        }

        if (!debug) exit(0);
    }


    if (snpname.compare("") == 0)
    {
        cerr << "Error: No SNP name given" << endl << endl;
        info(program_name);
        exit(2);
    }

    unsigned long int snprow = 0;
    bool snpfound = false;
    bool idfound  = false;


    // Look at the SNPs (rows) first
    if (debug)
    {
        cout << "---------- List of SNP names ----------" << endl;
    }
    for (unsigned long int row = 0; row < fv.getNumVariables(); row++)
    {
        string current_snpname = fv.readVariableName(row).name;
        if (current_snpname.compare(snpname) == 0)
        {
            snpfound = true;
            snprow = row;
            if (debug)
            {
                cout << "*" << current_snpname << "* ";
            }
        } else {
            if (debug)
            {
                cout << current_snpname << " ";
            }
        }
    }
    if (debug)
    {
        cout << endl;
        cout << "---------- End of SNP names ----------" << endl;

        cout << "N_obs = " << fv.getNumObservations() << "\telement size: "
             <<  fv.getElementSize() << "\tDataType: "
             << dataTypeToString(fv.getElementType()) << endl;
    }

    if (!snpfound)
    {
        cerr << "SNP name '" << snpname << "' not found in data file "
             << inputFileName << endl;
        exit(1);
    }

    char * data = new (nothrow) char[fv.getNumObservations() *
                                         fv.getElementSize()];
    if (!data)
    {
        cerr << "Cannot allocate memory for data vector. Exiting..." << endl;
        exit(2);
    }

    fv.readVariable(snprow, data);
    for (unsigned long int col = 0; col < fv.getNumObservations(); col++)
    {
        if (idname.compare("") != 0)
        {
            // An ID name has been set, only print the dosage for that ID
            // and SNP combination
            if (idname.compare(fv.readObservationName(col).name) == 0)
            {
                idfound = true;
                cout << fv.readObservationName(col).name << "\t"
                     << bufToString(fv.getElementType(),
                                    &data[col*fv.getElementSize()],
                                    string("NA"))
                     << endl;
            }
        } else {
            // Print the dosages for all IDs
            cout << fv.readObservationName(col).name << "\t"
                 << bufToString(fv.getElementType(),
                                &data[col*fv.getElementSize()],
                                string("NA"))
                 << endl;
        }
    }

    if (idfound == false && idname.compare("") != 0)
    {
        cerr << "Id '" << idname << "' not found in data file "
             << inputFileName << endl;
        exit(1);
    }

    delete [] data;
}
